stan/math/rev/functor/dae.hpp

```diff
@@ -0,0 +1,175 @@
+#ifndef STAN_MATH_REV_FUNCTOR_DAE_HPP
+#define STAN_MATH_REV_FUNCTOR_DAE_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/functor/idas_integrator.hpp>
+#include <stan/math/rev/functor/dae_system.hpp>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Solve the DAE initial value problem f(t, y, y')=0, y(t0) = yy0, y'(t0)=yp0 at
+ * a set of times, { t1, t2, t3, ... } using IDAS.
+ *
```

Comment:
IDAS is for solving DAEs of index-1, as it's based on the DASPK package (see https://epubs.siam.org/doi/abs/10.1137/0915088?journalCode=sjoce3). Basically is solves DAE of the form
```stan
y'=f(t, y)
0=g(t,y)
```
In the current implementation the user is responsible for providing consistent initial conditions. In this case one doesn't need to explicitly specify the differential & algebraic components. I plan to add an overload in next step to allow calculate such conditions, in that case one will need to tell the solver which states are algebraic only.

Reply:
So with the current prototype, the user needs to pass a residual function which has the form
```
h(t, y, y') = 0
```
where `h(t, y, y') = (y' - f(t, y), g(t, y))`. What happens if we pass a system which violates the above condition (say `h = (2 * y' - f, g)`)? Do we get an error message or just an incorrect result?

The problem is that we have no way of checking whether the user is violating this condition. If we forced the user to pass in both the LHS of `y'` (so `f`) and the algebraic components, we could circumvent this problem. But this means the user has to pass two functions; this complicates the function and prevents us from using variadic arguments.

> I plan to add an overload in next step to allow calculate such conditions, in that case one will need to tell the solver which states are algebraic only.

Yeah, I can envision a solution where the user specifies which states are algebraic.

I think we need to address this issue. But I open to hearing another opinion on the matter.

Reply:
`h=(2*y'-f, g)` is allowed. I should have been more specific. By "basically it solves DAE of the form..." I meant to say this is the type of index-1 DAEs. What IDAS does is to replace the `y'` term in the residual function `h` with its BDF discretization and solve an algebraic system (See Eq. 2.4 of IDAS user guide). So when user supplies `h(y', y, t)`, IDAS solves `h((y(n+1) - y(n))/2, y(n+1), t)=0` for unkown `y(n+1)` at step `n+1` (here we use backward Euler as an example). In this case user doesn't have to specify which eqn's are for ODE and which are for algebraic components.

Reply:
Thanks for the clarification. So there is no hard constraint on `h`? It could be `h = (2 * y', g)` or even `h = (y'^2, g)`, etc, and the solver will run? Sure, the user is responsible for writing a solvable algebraic constraint -- something we cannot check. But then again this is no different from a regular algebraic equation.

If my understanding is correct on this point, I'd say this resolves the issue.

Reply:
One hard constraint on `h`: it is of index-1, i.e. it takes one time-derivative operation & additional algebraic operations if necessary to turn `h=0` into an ODE `y'=f(y,t)`. See [this explanation](https://www.mathworks.com/help/matlab/math/solve-differential-algebraic-equations-daes.html). The examples you raised are index-1 DAEs so yes they can be solved by IDAS.

Reply:
Ok, I see. I'll admit I wasn't familiar with the differential index concept. What happens if the user passes an index-2 DAE? What are the mode of failures of the DAE in this case? Does IDAS produce an error message?

The documentation you wrote in the docs PR looks good. In addition to the reference, I would add a brief definition of index-1 and the link to the MathWorks page. I think it's also good to mention that higher index DAEs can be rewritten as index-1 DAEs.

Reply:
> What happens if the user passes an index-2 DAE? What are the mode of failures of the DAE in this case? Does IDAS produce an error message?

In general it will most likely fail with "Error test failure" or "Convergence failure". I just added a unit test for an index-3 DAE. It fails with error msg "Error test failures occurred too many times". Unfortunately there's no way to narrow down from such a msg to the exact cause so that we can remind users to perform some index reduction beforehand. It has to be user responsibility. 

I'll add more details in the doc PR for this.

Reply:
Ok. I think the most important here is that it fails with an error message.